lib/spack/spack/test/config.py

```diff
@@ -358,28 +358,51 @@ def test_substitute_tempdir(mock_low_high_config):
     )
 
 
-def test_substitute_padding(mock_low_high_config):
-    max_system_path = spack_path.get_system_path_max()
-    expected_length = (max_system_path -
-                       spack_path.SPACK_MAX_INSTALL_PATH_LENGTH)
-
-    install_path = spack_path.canonicalize_path('/foo/bar/${padding}/baz')
-
-    assert spack_path.SPACK_PATH_PADDING_CHARS in install_path
-    assert len(install_path) == expected_length
-
-    install_path = spack_path.canonicalize_path('/foo/bar/baz/gah/$padding')
-
-    assert spack_path.SPACK_PATH_PADDING_CHARS in install_path
-    assert len(install_path) == expected_length
-
-    i_path = spack_path.canonicalize_path('/foo/$padding:10')
-    i_expect = os.path.join('/foo', spack_path.SPACK_PATH_PADDING_CHARS[:10])
-    assert i_path == i_expect
-
-    i_path = spack_path.canonicalize_path('/foo/${padding:20}')
-    i_expect = os.path.join('/foo', spack_path.SPACK_PATH_PADDING_CHARS[:20])
-    assert i_path == i_expect
+PAD_STRING = spack.util.path.SPACK_PATH_PADDING_CHARS
+MAX_PATH_LEN = spack.util.path.get_system_path_max()
+MAX_PADDED_LEN = MAX_PATH_LEN - spack.util.path.SPACK_MAX_INSTALL_PATH_LENGTH
+reps = [PAD_STRING for _ in range((MAX_PADDED_LEN // len(PAD_STRING) + 1) + 2)]
+full_padded_string = os.path.join(
+    '/path', os.path.sep.join(reps))[:MAX_PADDED_LEN]
+
+
+@pytest.mark.parametrize('config_settings,expected', [
+    ([], [None, None, None]),
+    ([['config:install_tree:root', '/path']], ['/path', None, None]),
+    ([['config:install_tree', '/path']], ['/path', None, None]),
+    ([['config:install_tree:projections', {'all': '{name}'}]],
+     [None, None, {'all': '{name}'}]),
+    ([['config:install_path_scheme', '{name}']],
+     [None, None, {'all': '{name}'}]),
+    ([['config:install_tree:root', '/path'],
+      ['config:install_tree:padded_length', 11]],
+     [os.path.join('/path', PAD_STRING[:5]), '/path', None]),
+    ([['config:install_tree:root', '/path/$padding:11']],
+     [os.path.join('/path', PAD_STRING[:5]), '/path', None]),
+    ([['config:install_tree', '/path/${padding:11}']],
+     [os.path.join('/path', PAD_STRING[:5]), '/path', None]),
+    ([['config:install_tree:padded_length', False]], [None, None, None]),
+    ([['config:install_tree:padded_length', True],
+      ['config:install_tree:root', '/path']],
+     [full_padded_string, '/path', None]),
+    ([['config:install_tree:', '/path$padding']],
+     [full_padded_string, '/path', None]),
+    ([['config:install_tree:', '/path/${padding}']],
+     [full_padded_string, '/path', None]),
+])
+def test_parse_install_tree(config_settings, expected, mutable_config):
```

Comment:
I think comment explaining `config_settings` would be useful (each element is a 2-item list specifying a key-value pair that should be updated in the configuration).